### PR TITLE
Integrate global HD index into SQL engine

### PR DESCRIPTION
**Pull request by petrpleshachkov**
- On the query plan construction step select index for full scans
- Plan creation unit testing

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1488/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1488/Hazelcast-pr-builder-1488.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DataSerializableConventionsTest.test_dataSerializableClasses_areIdentifiedDataSerializable:121 There are 63 classes which are DataSerializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[ERROR]   DataSerializableConventionsTest.test_serializableClasses_areIdentifiedDataSerializable:161 There are 56 classes which are Serializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[INFO] 
[ERROR] Tests run: 3854, Failures: 2, Errors: 0, Skipped: 14
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1489/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1489/Hazelcast-pr-builder-1489.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DataSerializableConventionsTest.test_dataSerializableClasses_areIdentifiedDataSerializable:121 There are 63 classes which are DataSerializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[ERROR]   DataSerializableConventionsTest.test_serializableClasses_areIdentifiedDataSerializable:161 There are 56 classes which are Serializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[INFO] 
[ERROR] Tests run: 3854, Failures: 2, Errors: 0, Skipped: 14
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1490/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1490/Hazelcast-pr-builder-1490.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DataSerializableConventionsTest.test_dataSerializableClasses_areIdentifiedDataSerializable:121 There are 63 classes which are DataSerializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[ERROR]   DataSerializableConventionsTest.test_serializableClasses_areIdentifiedDataSerializable:161 There are 56 classes which are Serializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[INFO] 
[ERROR] Tests run: 3854, Failures: 2, Errors: 0, Skipped: 14
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

File: hazelcast-sql/pom.xml [hunk 1]
```diff
@@ -73,6 +73,20 @@
                     </execution>
                 </executions>
             </plugin>
+
+            <plugin>
```
Comment (devozerov) [inline hazelcast-sql/pom.xml:77]: What is the reason of this change? It is not needed for compilation in IDE. Is it required by Jenkins?

File: hazelcast/src/main/java/com/hazelcast/sql/impl/connector/LocalReplicatedMapConnector.java [hunk 2]
```diff
@@ -78,6 +78,8 @@ public Table createTable(
 
         long estimatedRowCount = stores.size() * nodeEngine.getPartitionService().getPartitionCount();
 
+        boolean nativeMemoryEnabled = nodeEngine.getConfig().getNativeMemoryConfig().isEnabled();
```
Comment (devozerov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/connector/LocalReplicatedMapConnector.java:81]: Do we support HD for replicated maps? I thought no.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/connector/LocalPartitionedMapConnector.java [hunk 3]
```diff
@@ -85,6 +85,7 @@ public Table createTable(
         long estimatedRowCount = estimatePartitionedMapRowCount(nodeEngine, context, mapName);
         List<MapTableIndex> indexes = container != null ? getPartitionedMapIndexes(container, fields) : emptyList();
         int distributionFieldOrdinal = container != null ? getPartitionedMapDistributionField(container, context, fields) : -1;
+        boolean nativeMemoryEnabled = nodeEngine.getConfig().getNativeMemoryConfig().isEnabled();
```
Comment (devozerov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/connector/LocalPartitionedMapConnector.java:88]: This flag looks suspicious to me. Even though an instance may have native memory enabled, this doesn't mean that the concrete map use NATIVE format.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/AbstractMapTable.java [hunk 4]
```diff
@@ -37,6 +37,7 @@
     private final UpsertTargetDescriptor keyUpsertDescriptor;
     private final UpsertTargetDescriptor valueUpsertDescriptor;
     private final QueryException exception;
+    private final boolean nativeMemoryEnabled;
```
Comment (devozerov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/AbstractMapTable.java:40]: I think this flag should be in the `PartitionedMapTable`, because replicated maps do not support indexes, and hence can never be used with HD. Old logic already handles this properly.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/ReplicatedMapTableResolver.java [hunk 5]
```diff
@@ -96,6 +96,8 @@ private ReplicatedMapTable createTable(ReplicatedMapService mapService, String n
 
                 long estimatedRowCount = stores.size() * nodeEngine.getPartitionService().getPartitionCount();
 
+                boolean nativeMemoryEnabled = nodeEngine.getConfig().getNativeMemoryConfig().isEnabled();
```
Comment (devozerov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/ReplicatedMapTableResolver.java:99]: Same as in other places: this is not needed for replicated maps

File: hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/visitor/PlanCreateVisitor.java [hunk 6]
```diff
@@ -277,9 +277,15 @@ private static SqlRowMetadata createRowMetadata(List<String> columnNames, List<Q
 
     @Override
     public void onMapScan(MapScanPhysicalRel rel) {
-        HazelcastTable hazelcastTable = rel.getTableUnwrapped();
         AbstractMapTable table = rel.getMap();
 
+        if (table.nativeMemoryEnabled()) {
+            throw QueryException.error("IMap scan with InMemoryFormat.NATIVE is not supported: " + table.getName()
```
Comment (devozerov) [inline hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/visitor/PlanCreateVisitor.java:283]: The error message is not very intuitive, because from the user perspective there is no "scan", but instead, there is an access to a table. I would change the message to something like this:
```
Cannot query IMap <name> with InMemoryFormat.NATIVE because it does not have indexes (please add at least one index to query this IMap)
```

File: hazelcast-sql/src/test/java/com/hazelcast/sql/impl/calcite/opt/OptimizerTestSupport.java [hunk 7]
```diff
@@ -155,22 +155,24 @@ private static PhysicalRel optimizePhysicalInternal(
     }
 
     protected static HazelcastTable partitionedTable(
```
Comment (devozerov) [inline hazelcast-sql/src/test/java/com/hazelcast/sql/impl/calcite/opt/OptimizerTestSupport.java:157]: I would propose to have two overloads, to avoid complication for non-HD tests (and the majority of tests are non-HD):
```
partitionedTable(name, fields, indexes, rowCount)
partitionedTable(name, fields, indexes, rowCount, nativeMemoryEnabled)
```

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java [hunk 8]
```diff
@@ -130,6 +124,8 @@ private PartitionedMapTable createTable(
             int distributionFieldOrdinal =
                 MapTableUtils.getPartitionedMapDistributionField(mapContainer, context, resolved.getFields());
 
+            boolean nativeMemoryEnabled = context.getNodeEngine().getConfig().getNativeMemoryConfig().isEnabled();
```
Comment (devozerov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java:127]: Same problem as in other places: this IMap is not necessarily HD map.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/connector/LocalReplicatedMapConnector.java (hunk 2 above)
Comment (pveentjer) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/connector/LocalReplicatedMapConnector.java:81]: no HD support for replicated maps

File: hazelcast/src/main/java/com/hazelcast/sql/impl/connector/LocalPartitionedMapConnector.java (hunk 3 above)
Comment (petrpleshachkov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/connector/LocalPartitionedMapConnector.java:88]: My impression was that once it is enabled, it is used for all maps. But, OK, if we still for some maps can use on-heap format, I'll initialize it on the map level.

However, here we don't know which format is used, because we might have `container = null`. I also see a comment "TODO: do we want to keep it? maps are auto discovered...".  So, the question here, do we have to pass nativeMemoryEnabled from the caller?

Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/connector/LocalPartitionedMapConnector.java:88]: If `container == null`, then the map cannot be used. We should not focus a lot on this specific part of the code too much, because `LocalPartitinedMapConnector` is research code. The following pseudocode should be sufficient:
```
nativeMemoryEnabled = container != null && nodeEngine.nativeMemoryEnabled() && containter.format == NATIVE
```

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1513/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1513/Hazelcast-pr-builder-1513.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast-sql: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/visitor/PlanCreateVisitor.java:284: Line is longer than 130 characters (found 195). [LineLength]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1515/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1515/Hazelcast-pr-builder-1515.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DataSerializableConventionsTest.test_dataSerializableClasses_areIdentifiedDataSerializable:121 There are 63 classes which are DataSerializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[ERROR]   DataSerializableConventionsTest.test_serializableClasses_areIdentifiedDataSerializable:161 There are 56 classes which are Serializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[INFO] 
[ERROR] Tests run: 3854, Failures: 2, Errors: 0, Skipped: 14
[INFO] 

[ERROR] There are test failures.
</details>
</pre>
